gravitee-io/gravitee-access-management
# feat: selection rules for application identity providers

fixes: https://github.com/gravitee-io/issues/issues/5167

⚠️ Do not merge before this is merged : 
- https://github.com/gravitee-io/issues/issues/7193
- https://github.com/gravitee-io/issues/issues/6519
# How to test:

## Selection Rules

Go to your application and go to the `Identity Provider` section
Click on the Rule icon.

If the rule is empty, it will suggest a rule, otherwise it will show the rule already present

<img width="1576" alt="Screenshot 2022-02-23 at 21 14 55" src="https://user-images.githubusercontent.com/8531515/155400837-34a9ca4e-8abb-4209-a957-ae0cb68389c2.png">

<img width="1575" alt="Screenshot 2022-02-23 at 21 14 32" src="https://user-images.githubusercontent.com/8531515/155400797-b66fdaa8-078b-402b-9d0d-ce59de7ba160.png">

Associate 

1. For Internal providers
- If the rule is null or empty, the provider will be included, this is to preserve previous behaviour.
- If the rule is correctly formatted, it will filter regarding the result of the rule

To test:
- Create several internal providers and add some rules
- Go to the login flow 
- Change the rules to see if it matches

2. For social providers
- In the `Settings > Login` section, enable `Identifier First Login`
- Go to the login flow and login with an email

## Identifier First Flows (Policies)

You can add policies within the Flows for ID first login.
In order to test, add a policy and include the created value within a Identity Provider Rule

<img width="1763" alt="Screenshot 2022-02-23 at 21 20 05" src="https://user-images.githubusercontent.com/8531515/155401581-1b65f938-3491-4250-868e-2c1180b1221b.png">

## Domain Whitelists

Domain whitelist have been introduced in order to select the right social / enterprise provider base on the email domain source.
Now they are used as a `/login/callback` post filter. If the email of the authenticated user does not match the domain. The use will be logged out.

To test:
- Use the Social login flow test above and input the domain you wish. If it matches you are logged in, otherwise you are sent back to the identifier first login

⚠️ This is a breaking change and it must be documented

<img width="1358" alt="image" src="https://user-images.githubusercontent.com/8531515/155402858-85432f64-b85d-4bfc-833e-b3ab8d78c5d7.png">
